plugins/modules/ec2_vpc_endpoint_service.py

```diff
@@ -0,0 +1,513 @@
+            set(update_endpoint_configuration[f'remove_{field}'])
+        )
+
+        if len(update_endpoint_configuration[f'remove_{field}']) > 0 or len(update_endpoint_configuration[f'add_{field}']) > 0:
+            update_endpoint_configuration['update_needed'] = True
+
+    return update_endpoint_configuration
+
+
+def get_vpc_endpoint_service_permissions(module: AnsibleAWSModule, client, service_id: str) -> list:
+
+    permissions_paginator = client.get_paginator('describe_vpc_endpoint_service_permissions')
+    permissions_principals = []
+    for page in permissions_paginator.paginate(ServiceId=service_id):
+        permissions_principals.extend(page['AllowedPrincipals'])
+
+    permissions = list(map(lambda AllowedPrincipal: AllowedPrincipal.get('Principal'), permissions_principals))
+    if permissions is None:
+        permissions = []
+    return permissions
+
+
+def vpc_endpoint_service_permissions_update_needed(service_id: str, existing_permissions: list, new_permissions: list, purge_permissions: bool) -> dict:
+
+    update_permissions = {
+        'update_needed': False,
+        'service_id': service_id
+    }
+
+    if purge_permissions:
+        update_permissions['remove_allowed_principals'] = list(set(existing_permissions) - set(new_permissions))
+    else:
+        update_permissions['remove_allowed_principals'] = []
+
+    update_permissions['add_allowed_principals'] = list(set(existing_permissions).symmetric_difference(set(new_permissions)) - set(existing_permissions))
+
+    if len(update_permissions['remove_allowed_principals']) > 0 or len(update_permissions['add_allowed_principals']) > 0:
+        update_permissions['update_needed'] = True
+
+    return update_permissions
+
+
+def modify_vpc_endpoint_service_permissions(module: AnsibleAWSModule, client, permission_params: list) -> dict:
+
+    try:
+        response = client.modify_vpc_endpoint_service_permissions(**permission_params)
+    except (
+        BotoCoreError,
+        ClientError,
+    ) as err:
+        module.fail_json_aws(err, msg="Failed to modify permissions {err}")
+    return response
+
+
+def format_check_mode_response(endpoint_service: dict, tags: dict, delete: bool = False) -> dict:
+
+    timestamp = datetime.now().isoformat()
+    if delete:
+        return {
+            "endpoint_service": endpoint_service,
+            "deletion_date": timestamp,
+            "version_id": "",
+        }
+    else:
+        return {
+            "creation_date": timestamp,
+            "version_id": "",
+            "endpoint_service": dict({'tags': tags}, **endpoint_service),
+        }
+
+
+def main():
+    module = AnsibleAWSModule(
+        argument_spec=ARGUMENT_SPEC,
+        required_if=REQUIRED_IF,
+        supports_check_mode=SUPPORTS_CHECK_MODE,
+    )
+
+    # Set initial result values
+    result = dict(changed=False, exists=False)
+    client = module.client("ec2")
```

Comment:
It would be good to use the AWSRetry decorator here
https://github.com/ansible-collections/amazon.aws/blob/main/docs/docsite/rst/dev_guidelines.rst#api-throttling-rate-limiting-and-pagination

```suggestion
    client = module.client("ec2", retry_decorator=AWSRetry.jittered_backoff())
```

This then means that (almost) all boto3 calls can have automatic retries applied to them just by passing "aws_retry=True" in addition the the usual arguments.